flink-tests/src/test/java/org/apache/flink/test/classloading/jar/CheckpointedStreamingProgram.java

```diff
@@ -72,34 +74,27 @@ public void run(SourceContext<String> ctx) throws Exception {
 		public void cancel() {
 			running = false;
 		}
-
-		@Override
-		public Integer snapshotState(long checkpointId, long checkpointTimestamp) throws Exception {
-			return null;
-		}
-
-		@Override
-		public void restoreState(Integer state) {
-
-		}
 	}
 
-	public static class StatefulMapper implements MapFunction<String, String>, Checkpointed<StatefulMapper>, CheckpointListener {
+	public static class StatefulMapper implements MapFunction<String, String>, ListCheckpointed<StatefulMapper>, CheckpointListener {
 
 		private String someState;
 		private boolean atLeastOneSnapshotComplete = false;
 		private boolean restored = false;
 
 		@Override
-		public StatefulMapper snapshotState(long checkpointId, long checkpointTimestamp) throws Exception {
-			return this;
+		public List<StatefulMapper> snapshotState(long checkpointId, long timestamp) throws Exception {
+			return Collections.singletonList(this);
 		}
 
 		@Override
-		public void restoreState(StatefulMapper state) {
-			restored = true;
-			this.someState = state.someState;
-			this.atLeastOneSnapshotComplete = state.atLeastOneSnapshotComplete;
+		public void restoreState(List<StatefulMapper> state) throws Exception {
+			if (!state.isEmpty()) {
```

Comment:
Alright, i figured out why we can't fail here immediately. It still seems odd though that do not explicitly differentiate between a call to restore before any state was snapshotted and a broken snapshotting that doesn't return a state, although this applies to all other tests as well.

If the test is successful f that we are getting the state that we snapshotted we should also have failure condition in case this does not happen; currently we simply enter undefined territory.